### ISSUE TITLE
How do multiple point detection works?

### ISSUE THREAD

**Issue by nviwch** [Jeff-sjtu/res-loglikelihood-regression#22]
I am confused how multiple key point detection works. One-point is easy, just compare the result. Two-point is not, because the model could output A, B or B, A, while the ground truth is a, b. So, multiple-point can have different permutation to compare. 

Another thing is, it is possible the model output shape is (B, N, 2) but the number of points needed to be detected is smaller than N. Then, how do you know the exact number of points dynamically.  

How about model output repeated points? NMS?

**Comment by jeffffffli** [comment]
Hi @nviwch,

If you mean how to conduct multi-person pose estimation, we follow the top-down framework. We first do object detection to find all the human bounding boxes. Then we assume each bounding box only contains one person and conduct single-person pose estiamtion within the cropped box. This process is just like the heatmap-based method. The redundant outputs are eliminated by NMS.

**Comment by nviwch** [comment]
Because I am trying to merge the DETR loss function with your flow model, they used Hungarian algorithm to get the argmin of permutation.

I am asking how do you make sure the order of object detection prediction and  ground truth are matched. For example, if ground truth is [(1,2), (3,4)] but model prediction is [(3.1,3.9), (1.1,1.9)], the best permutation is obviously (3.1, 3.9) <-> (3,4) & (1.1,1.9) <-> (1,2) and the loss is smaller than comparing with the other mapping.

**Comment by jeffffffli** [comment]
I think you can directly use the Hungarian matching like DETR. First, you can use conventional Euclidean distance to find the best match. Then use the match results to calculate the RLE loss.

Additionally, we found that to make RLE loss works well, the coordinates should be normalized to the same scale.

**Comment by nviwch** [comment]
Thanks. I have written my set based loss based on DETR with ordinary CNN (no transformer) with simple task in tensorflow. Locate variable number of shapes and classify their classes at the same time. The power of the loss is I do not need to apply NMS as post processing.

![29](https://user-images.githubusercontent.com/19923010/139178800-10b5d4b6-e99f-473f-aacc-ffef803286fd.png) ![74](https://user-images.githubusercontent.com/19923010/139178822-8f5b687f-d36e-4198-93f4-a4538c6eaa32.png) ![538](https://user-images.githubusercontent.com/19923010/139179026-f4e6044e-2a91-459f-90da-145cabce3f21.png)

Writing set based loss with RLE.

**Comment by jeffffffli** [comment]
That's a pretty cool result. 🆒

> Thanks. I have written my set based loss based on DETR with ordinary CNN (no transformer) with simple task in tensorflow. Locate variable number of shapes and classify their classes at the same time. The power of the loss is I do not need to apply NMS as post processing.
> 
> ![29](https://user-images.githubusercontent.com/19923010/139178800-10b5d4b6-e99f-473f-aacc-ffef803286fd.png) ![74](https://user-images.githubusercontent.com/19923010/139178822-8f5b687f-d36e-4198-93f4-a4538c6eaa32.png) ![538](https://user-images.githubusercontent.com/19923010/139179026-f4e6044e-2a91-459f-90da-145cabce3f21.png)
> 
> Writing set based loss with RLE.

**Comment by nviwch** [comment]
I have integrate the RLE flow model with the set based loss. It could represents some difficult or low confidence shapes with higher sigma

The below colour scheme is different with before. colour = (1 - no_obj_prob) * (0, 255 * circle_prob, 255 * sq_prob), while the x_width = 1 + sigma_x * 5, y_width = 1 + sigma_y * 5

![181](https://user-images.githubusercontent.com/19923010/140001813-60ea7194-48ae-4460-9c1c-bd7bb7025459.png) ![321](https://user-images.githubusercontent.com/19923010/140001861-0531cc84-49ff-4026-a20d-d4473df161a4.png) ![587](https://user-images.githubusercontent.com/19923010/140001912-0f059112-31c1-43a1-83f1-63406b6dd295.png)

**Comment by jeffffffli** [comment]
> I have integrate the RLE flow model with the set based loss. It could represents some difficult or low confidence shapes with higher sigma
> 
> The below colour scheme is different with before. colour = (1 - no_obj_prob) * (0, 255 * circle_prob, 255 * sq_prob), while the x_width = 1 + sigma_x * 5, y_width = 1 + sigma_y * 5
> 
> ![181](https://user-images.githubusercontent.com/19923010/140001813-60ea7194-48ae-4460-9c1c-bd7bb7025459.png) ![321](https://user-images.githubusercontent.com/19923010/140001861-0531cc84-49ff-4026-a20d-d4473df161a4.png) ![587](https://user-images.githubusercontent.com/19923010/140001912-0f059112-31c1-43a1-83f1-63406b6dd295.png)

I am glad to see RLE can generalize well in different tasks. It seems to be more robust in challenging scenes than conventional regression.
